staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/generic/controller.go

```diff
@@ -0,0 +1,255 @@
+	//!TODO: check if informer is even valid?
+	// e.g. if crd isnt even installed yet this just waits forever here
+	if !cache.WaitForNamedCacheSync(c.options.Name, ctx.Done(), c.informer.HasSynced) {
+		// ctx cancelled during cache sync. return early
+		err := ctx.Err()
+		if err == nil {
+			// if context wasnt cancelled then the sync failed for another reason
+			err = errors.New("cache sync failed")
+		}
+		return err
+	}
+
+	waitGroup := sync.WaitGroup{}
+
+	for i := uint(0); i < c.options.Workers; i++ {
+		waitGroup.Add(1)
+		go func() {
+			wait.Until(c.runWorker, time.Second, ctx.Done())
+			waitGroup.Done()
+		}()
+	}
+
+	klog.Infof("Started %v workers for %v", c.options.Workers, c.options.Name)
+
+	// Wait for context cancel.
+	<-ctx.Done()
+
+	// Gracefully shutdown workqueue. Finish processing any enqueued items.
+	//!TODO: May want to put deadline on this and forcefully shutdown?
+	c.queue.ShutDownWithDrain()
+
+	// Workqueue shutdown signals for workers to stop. Wait for all workers to
+	// clean up
+	waitGroup.Wait()
+
+	// Remove event handler and ignore error (only raised for improper usage)
+	c.informer.RemoveEventHandler(registration)
+
+	// Only way for workers to ever stop is for caller to cancel the context
+	return ctx.Err()
+}
+
+func (c *controller[T]) HasSynced() bool {
+	return c.informer.HasSynced()
+}
+
+func (c *controller[T]) runWorker() {
+	for {
+		obj, shutdown := c.queue.Get()
+		if shutdown {
+			return
+		}
+
+		// We wrap this block in a func so we can defer c.workqueue.Done.
+		err := func(obj interface{}) error {
+			// We call Done here so the workqueue knows we have finished
+			// processing this item. We also must remember to call Forget if we
+			// do not want this work item being re-queued. For example, we do
+			// not call Forget if a transient error occurs, instead the item is
+			// put back on the workqueue and attempted again after a back-off
+			// period.
+			defer c.queue.Done(obj)
+			var key string
+			var ok bool
+			// We expect strings to come off the workqueue. These are of the
+			// form namespace/name. We do this as the delayed nature of the
+			// workqueue means the items in the informer cache may actually be
+			// more up to date that when the item was initially put onto the
+			// workqueue.
+			if key, ok = obj.(string); !ok {
+				// As the item in the workqueue is actually invalid, we call
+				// Forget here else we'd go into a loop of attempting to
+				// process a work item that is invalid.
+				c.queue.Forget(obj)
+				return fmt.Errorf("expected string in workqueue but got %#v", obj)
+			}
+
+			if err := c.reconcile(key); err != nil {
+				// Put the item back on the workqueue to handle any transient errors.
+				c.queue.AddRateLimited(key)
+				return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+			}
+			// Finally, if no error occurs we Forget this item so it does not
+			// get queued again until another change happens.
+			c.queue.Forget(obj)
```

Comment:
You should not forget an item in this situation if I understand what is going on, just use Done() (already deferred). It might have been dirtied since you started working.

Reply:
It is my understanding that `Forget` resets the rate limiter for that item. Isn't it necessary to call this upon success otherwise a once-rate-limited object might forever be rate limited?

FWIW most of this worker code is adapted from `sample-controller` which has the same behavior:

https://github.com/kubernetes/kubernetes/blob/b62774f2f7907e0d519026e8fc3cfaf50eaa5d31/staging/src/k8s.io/sample-controller/controller.go#L210-L229

Reply:
Ah, I expected Forget to remove it from the delayed queue, and judging from the comment, so did whoever wrote this originally :)

Reply:
Welp I've corrected the comments here and made a note to myself to go to sample controller and fix it up there. Marking this as resolved here